Feature/m0228 j 473 update kifaru doc (#343)
MicroEJ/docs: ApplicationDeveloperGuide/js/java.rst

The MicroEJ engine allows to expose Java objects or methods to the JavaScript code by using the engine API and creating the adequate JavaScript object.

> Java objects can be exposed to JavaScript using the ``JavaImport`` mechanism: 

Reviewer: ```suggestion
Java objects can be exposed to JavaScript using the ``JavaImport`` mechanism. 
```

Author: fixed

> it takes a Java fully qualified name as argument and returns an objects that 

Reviewer: ```suggestion
It takes a Java fully qualified name as argument and returns an object that 
```

Author: fixed

> gives access to the constructors, static methoods and static fields.

Reviewer: ```suggestion
gives access to the constructors, static methods and static fields.
```

> gives access to the constructors, static methoods and static fields.

Reviewer: Maybe we should add a sentence to clearly state which Java classes can be imported (classes of the application and from the application dependencies).

Author: Done

For instance, the following code imports ``java.lang.System`` and prints a 
string calling ``System.out.println()``:

.. code-block:: javascript

	var System = JavaImport("java.lang.System")
	System.out.println("foo");

> Here we instantiate a Java ``File`` object and checks that it exists:

Reviewer: ```suggestion
Here we instantiate a Java ``File`` object and check that it exists:
```

Author: fixed

.. code-block:: javascript

	var File = JavaImport("java.io.File")
	var myFile = new File("myFile.txt")

	if (myFile.exists()) {
		print("myFile.txt exists")
	} else {
		print("myFile.txt does not exist")
	}

.. warning::

     You cannot instantiate an anonymous class from an interface or an abstract class with the ``new`` keyword and ``JavaImport``. Nevertheless, you can still access to static fields and methods.   

> We can also implement JavaScript functions in Java by adding their 

Reviewer: Shoudn't we split in several sections, just to distinguish clearly the different ways to call Java from JavaScript ?

Author: Done

.. code-block:: java

    JsObjectFunction functionObject = (JsObjectFunction) JsRuntime.JS_GLOBAL_OBJECT.get("sum");
    JsRuntime.ENGINE.addJob(functionObject, JsRuntime.JS_GLOBAL_OBJECT, new Integer(5), new Integer(3));

The first line gets the JavaScript function from the global scope.
The second line adds a job in the JavaScript engine queue to execute the function, in the global scope, with the arguments ``5`` and ``3``.

> Passing Values Between JavaScript and Java

Reviewer: ```suggestion
Passing values between JavaScript and Java
```

Author: Titles are capitalized, why not these two words?

Reviewer: I don't know what is the rule exactly, but it looks weird to have words like "between" capitalized.

------------------------------------------

JavaScript base types are represented by Java objects and not Java base types. 

> The following table show the correspondance between types in both languages: 

Reviewer: ```suggestion
The following table shows the mapping between types in both languages: 
```

Author: fixed

.. code-block:: java

    public class Callback<T> {

        @Nullable
        private T value;

        private boolean returned;

        /**
         * Gets the value returned by this callback function when ready.
         * <p>
         * A call to this method waits for the value to be ready.
         *
         * @return the value return by the callback
         */
	    @Nullable
        public T getValue() {
            synchronized (this) {
                while (!this.returned) {
                    try {
                        wait();
                    } catch (InterruptedException e) {
                        throw new JsErrorWrapper(""); //$NON-NLS-1$
                    }
                }
            }

             return this.value;
        }

        /**
         * Sets the value to return by this callback function.
         *
         * @param value
         *            the value to return
         */
        public synchronized void returnValue(@Nullable T value) {
            this.value = value;
            this.returned = true;
            notify();
        }
    }

We can now pass the callback to the job. The Java code will wait on the 

> ``callback.getValue()`` untill the result is ready.

Reviewer: ```suggestion
``callback.getValue()`` until the result is ready.
```

Author: fixed
